### PR TITLE
Issue 94 resolved with triggers

### PR DESCRIPTION
**Pull request by JoshuaTPritchett**
https://github.com/rht-labs/labs-ci-cd/issues/94

### CODE REVIEW

**Comment by sherl0cks** [comment]
@JoshuaTPritchett Thanks for contributing. See below the build config generated by your changes in the CI job. I think for this to work as expected, you'll need to declare a parameter for each of these secrets, and then reference the parameter name in the value of the template. Check out [this example](https://github.com/openshift/origin/blob/master/examples/db-templates/postgresql-persistent-template.json#L252) of generating the password for PostgreSQL

```
apiVersion: build.openshift.io/v1
kind: BuildConfig
metadata:
  creationTimestamp: '2018-01-29T03:55:15Z'
  labels:
    build: java-app
    name: java-app-pipeline
    template: s2i-app-build-template
    type: pipeline
  name: java-app-pipeline
  namespace: labs-ci-cd-pr-114
  resourceVersion: '6475648'
  selfLink: >-
    /apis/build.openshift.io/v1/namespaces/labs-ci-cd-pr-114/buildconfigs/java-app-pipeline
  uid: 36a2e7c7-04a8-11e8-9dd8-fa163eee5816
spec:
  nodeSelector: {}
  output: {}
  postCommit: {}
  resources: {}
  runPolicy: Serial
  source:
    git:
      ref: dev-demo
      uri: 'https://github.com/rht-labs/automation-api'
    type: Git
  strategy:
    jenkinsPipelineStrategy:
      jenkinsfilePath: Jenkinsfile
    type: JenkinsPipeline
  triggers:
    - type: ConfigChange
    - github:
        secret: '[A-Z0-9]{4}'
      type: GitHub
    - generic:
        secret: '[A-Z0-9]{4}'
      type: Generic
    - gitlab:
        secret: '[A-Z0-9]{4}'
      type: GitLab
    - bitbucket:
        secret: '[A-Z0-9]{4}'
      type: Bitbucket
status:
  lastVersion: 1
```

File: openshift-templates/s2i-app-build/template-no-secrets.json [hunk 1]
```diff
@@ -54,9 +54,33 @@
...
                 "triggers": [
-                    {
-                        "type": "ConfigChange"
+                  {
+                    "type": "ConfigChange"
+                  },
+                  {
+                    "type": "GitHub",
+                    "github": {
+                      "secret": "[A-Z0-9]{4}"
```
Comment (sherl0cks) [inline openshift-templates/s2i-app-build/template-no-secrets.json]: needs to be updated a la [this example](https://github.com/openshift/rails-ex/blob/master/openshift/templates/rails-postgresql.json#L510)

File: openshift-templates/s2i-app-build/template-no-secrets.json [hunk 2]
```diff
@@ -54,9 +54,33 @@
...
                     }
+                  },
+                  {
+                    "type": "Generic",
+                    "generic": {
+                      "secret": "[A-Z0-9]{4}"
```
Comment (sherl0cks) [inline openshift-templates/s2i-app-build/template-no-secrets.json]: needs to updated a la [this example](https://github.com/openshift/rails-ex/blob/master/openshift/templates/rails-postgresql.json#L510)

File: openshift-templates/s2i-app-build/template-no-secrets.json [hunk 3]
```diff
@@ -54,9 +54,33 @@
...
+                    }
+                  },
+                  {
+                    "type": "GitLab",
+                    "gitlab": {
+                      "secret": "[A-Z0-9]{4}"
```
Comment (sherl0cks) [inline openshift-templates/s2i-app-build/template-no-secrets.json]: needs to updated a la [this example](https://github.com/openshift/rails-ex/blob/master/openshift/templates/rails-postgresql.json#L510)

File: openshift-templates/s2i-app-build/template-no-secrets.json [hunk 4]
```diff
@@ -54,9 +54,33 @@
...
+                    }
+                  },
+                  {
+                    "type": "Bitbucket",
+                    "bitbucket": {
+                      "secret": "[A-Z0-9]{4}"
```
Comment (sherl0cks) [inline openshift-templates/s2i-app-build/template-no-secrets.json]: needs to updated a la [this example](https://github.com/openshift/rails-ex/blob/master/openshift/templates/rails-postgresql.json#L510)

File: openshift-templates/s2i-app-build/template-with-secrets.json [hunk 5]
```diff
@@ -58,9 +58,33 @@
...
                 "triggers": [
-                    {
-                        "type": "ConfigChange"
+                  {
+                    "type": "ConfigChange"
+                  },
+                  {
+                    "type": "GitHub",
+                    "github": {
+                      "secret": "[A-Z0-9]{4}"
```
Comment (sherl0cks) [inline openshift-templates/s2i-app-build/template-with-secrets.json]: needs to updated a la [this example](https://github.com/openshift/rails-ex/blob/master/openshift/templates/rails-postgresql.json#L510)

File: openshift-templates/s2i-app-build/template-with-secrets.json [hunk 6]
```diff
@@ -58,9 +58,33 @@
...
                     }
+                  },
+                  {
+                    "type": "Generic",
+                    "generic": {
+                      "secret": "[A-Z0-9]{4}"
```
Comment (sherl0cks) [inline openshift-templates/s2i-app-build/template-with-secrets.json]: needs to updated a la [this example](https://github.com/openshift/rails-ex/blob/master/openshift/templates/rails-postgresql.json#L510)

File: openshift-templates/s2i-app-build/template-with-secrets.json [hunk 7]
```diff
@@ -58,9 +58,33 @@
...
+                    }
+                  },
+                  {
+                    "type": "GitLab",
+                    "gitlab": {
+                      "secret": "[A-Z0-9]{4}"
```
Comment (sherl0cks) [inline openshift-templates/s2i-app-build/template-with-secrets.json]: needs to updated a la [this example](https://github.com/openshift/rails-ex/blob/master/openshift/templates/rails-postgresql.json#L510)

File: openshift-templates/s2i-app-build/template-with-secrets.json [hunk 8]
```diff
@@ -58,9 +58,33 @@
...
+                    }
+                  },
+                  {
+                    "type": "Bitbucket",
+                    "bitbucket": {
+                      "secret": "[A-Z0-9]{4}"
```
Comment (sherl0cks) [inline openshift-templates/s2i-app-build/template-with-secrets.json]: needs to updated a la [this example](https://github.com/openshift/rails-ex/blob/master/openshift/templates/rails-postgresql.json#L510)

**Comment by JoshuaTPritchett** [comment]
@sherl0cks Sorry, this a new construct that I just learned. The topic wasn't quite clear on the docs, thanks for the example, I hope this is what you were looking for. As a side note, I made the expression 16 character s long, I am not sure if this is too long or short?

**Comment by sherl0cks** [comment]
@JoshuaTPritchett no need to apologize - this is why we peer review changes. Your changes look good in CI. RE: length, it looks like `oc new-app` creates secrets of length `20`, so if you can update your secrets to that length, I think we are good to merge this in

**Comment by JoshuaTPritchett** [comment]
Done.

**Comment by sherl0cks** [comment]
@JoshuaTPritchett thank you. we welcome your contributions. feel free to keep pulling issues!

**Pull request merged**